src/AllocationIDTracker.sol

```diff
@@ -0,0 +1,51 @@
+// Copyright 2023-, Semiotic AI, Inc.
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.13;
+
+/**
+ * @title AllocationIDTracker
+ * @dev This contract tracks the allocation IDs of the RAVs that have been submitted to
+ *      ensure that each allocation ID is only used once. It is external to collateral
+ *      contract to allow for updating the collateral contract without losing the list of
+ *      used allocation IDs.
+ * @notice This contract is intended to be used with the `Collateral` contract.
+ */
+contract AllocationIDTracker {
+    mapping (address => bool) private _usedAllocationIDs;
+
+    /**
+     * @dev Emitted when an allocation ID is used.
+     */
+    event AllocationIDUsed(address indexed allocationID);
+
+    /**
+     * @dev Checks if an allocation ID has been used.
+     * @param allocationID The allocation ID to check.
+     * @return True if the allocation ID has been used, false otherwise.
+     */
+    function isAllocationIDUsed(address allocationID) public view returns (bool) {
+        return _usedAllocationIDs[allocationID];
+    }
+
+    /**
+     * @dev Marks an allocation ID as used.
+     * @param allocationID The allocation ID to mark as used.
+     * @notice REVERT: This function may revert if the allocation ID has already been used.
+     */
+    function useAllocationID(address allocationID) public {
```

Comment:
This being permissionless allows anyone to mark allocations as already used, potentially denying fees to everyone by front running txs. 


One way of protecting these methods is to require the caller to post a proof that they own the key pair behind an allocation id (see [this](https://github.com/613f824f0c3a43be9b4d07fa5e297ff0?pvs=25) snippet).